### PR TITLE
Extract root columns from BigQuery schema

### LINKED ISSUE #51: Top-level BigQuery schema should be a JSON list
https://cloud.google.com/bigquery/docs/schemas#creating_a_JSON_schema_file

### PR DESCRIPTION
**Pull request by acmiyaguchi**
This will fix #51 by making the starting the resulting BigQuery schema as a list of elements.

EDIT: This is ready for review. See https://github.com/mozilla/jsonschema-transpiler/pull/61#issuecomment-488427937

---
There are currently 5 failures:

```
failures:
    bigquery_test_object_empty_record
    bigquery_test_oneof_object_merge
    bigquery_test_oneof_object_merge_nullability
    bigquery_test_oneof_object_merge_with_complex
    bigquery_test_oneof_object_with_atomics

test result: FAILED. 23 passed; 5 failed; 0 ignored; 0 measured; 0 filtered out
```

In particular, the edge case that needs to be captured is the empty schema. Should the resulting table be an empty schema i.e. `[]`?

```json
    {
      "description": [
        "Empty structs are not supported in BigQuery, so we treat them ",
        "as empty documents."
      ],
      "name": "test_object_empty_record",
      "test": {
        "avro": {
          "type": "string"
        },
        "bigquery": {
          "mode": "REQUIRED",
          "type": "STRING"
        },
        "json": {
          "properties": {},
          "type": "object"
        }
      }
    }
```

I suspect there may need to be changes to the avro implementation too. However, all real-world documents will never reach this edge-case due to valid metadata.

### CODE REVIEW

**Comment by acmiyaguchi** [comment]
@jklukas Do you have any thoughts on what the default empty/incompatible schema be transformed into? With the JSON-BQ preprocessor, we could get away with throwing away the field and collecting the additional properties. This could also be an exceptional case that causes the program to panic -- this would force us to have at least one section in the schema (metadata) so that the schema will compile.

**Comment by jklukas** [comment]
I'm fine with the empty schema being an error that causes the program to panic. As you say, we will always be preprocessing to add metadata fields before sending a json schema through the transpiler, so if we run into this case, something has gone wrong and failing is probably better than letting some default slip through.

**Comment by fbertsch** [comment]
empty/incompatible schemas should be caught at schema creation time and not deploy time, so having the program panic would force the tests to fail in MPS, which is probably the desired behavior.

**Comment by acmiyaguchi** [comment]
I've actually been trying out a PoC for https://github.com/mozilla/jsonschema-transpiler/issues/63#issue-439234960 and have found some pretty useful things based on empirical tests. It looks like avro records with primitive types e.g. `{"type": "long"}` will be turned into `[{"mode":"REQUIRED","name":"root","type":"INTEGER"}]`. This means there there's no need to panic.

**Comment by acmiyaguchi** [comment]
I wrote a handy script to verify some of my findings. I think its a good approach, but it needs some integration into the rest of the workflow. I'm leaving it in a different branch to be worked on.

```bash
$ ./scripts/generate-bq-schema.sh 
4/4 succeeded
Copying file://avro-data/test-array.avro [Content-Type=application/octet-stream]...
Copying file://avro-data/test-long.avro [Content-Type=application/octet-stream]...
Copying file://avro-data/test-map.avro [Content-Type=application/octet-stream]...
Copying file://avro-data/test-string.avro [Content-Type=application/octet-stream]...
- [4 files][  420.0 B/  420.0 B]                                                
Operation completed over 4 objects/420.0 B.                                      
Dataset 'amiyaguchi-unmap-json-bq-sink:test_bq_schema' already exists.
Waiting on bqjob_r54b50f51158d3df5_0000016a75376f88_1 ... (2s) Current status: DONE   
original avro schema:
{
  "items": {
    "type": "long"
  },
  "type": "array"
}
bigquery schema:
[
  {
    "type": "INTEGER",
    "name": "root",
    "mode": "REPEATED"
  }
]
Waiting on bqjob_r6dceaf6af10fa2e6_0000016a75378767_1 ... (4s) Current status: DONE   
original avro schema:
{
  "type": "string"
}
bigquery schema:
[
  {
    "type": "STRING",
    "name": "root",
    "mode": "REQUIRED"
  }
]
Waiting on bqjob_r102e0e87e876af9c_0000016a7537a5fd_1 ... (3s) Current status: DONE   
original avro schema:
{
  "type": "long"
}
bigquery schema:
[
  {
    "type": "INTEGER",
    "name": "root",
    "mode": "REQUIRED"
  }
]
Waiting on bqjob_r16185d6871d7b40_0000016a7537c084_1 ... (2s) Current status: DONE   
original avro schema:
{
  "type": "map",
  "values": {
    "type": "long"
  }
}
bigquery schema:
[
  {
    "fields": [
      {
        "type": "STRING",
        "name": "key",
        "mode": "REQUIRED"
      },
      {
        "type": "INTEGER",
        "name": "value",
        "mode": "REQUIRED"
      }
    ],
    "type": "RECORD",
    "name": "root",
    "mode": "REPEATED"
  }
]
```

**Comment by fbertsch** [comment]
@acmiyaguchi EDIT: I was using the wrong binary. It worked!

```
fbertsch-23817:jsonschema-transpiler frankbertsch$ target/debug/jsonschema-transpiler core.10.schema.json --type bigquery
[
  {
    "mode": "REPEATED",
    "name": "accessibilityServices",
    "type": "STRING"
  },
  {
    "mode": "REQUIRED",
    "name": "arch",
    "type": "STRING"
  },
  {
    "mode": "NULLABLE",
    "name": "bug_1501329_affected",
    "type": "BOOL"
  },
  {
    "mode": "NULLABLE",
    "name": "campaign",
    "type": "STRING"
  },
  {
    "mode": "REQUIRED",
    "name": "clientId",
    "type": "STRING"
  },
  {
    "mode": "NULLABLE",
    "name": "created",
    "type": "STRING"
  },
  {
    "mode": "NULLABLE",
    "name": "defaultBrowser",
    "type": "BOOL"
  },
  {
    "mode": "NULLABLE",
    "name": "defaultSearch",
    "type": "STRING"
  },
  {
    "mode": "REQUIRED",
    "name": "device",
    "type": "STRING"
  },
  {
    "mode": "NULLABLE",
    "name": "displayVersion",
    "type": "STRING"
  },
  {
    "mode": "NULLABLE",
    "name": "distributionId",
    "type": "STRING"
  },
  {
    "mode": "NULLABLE",
    "name": "durations",
    "type": "INT64"
  },
  {
    "mode": "REPEATED",
    "name": "experiments",
    "type": "STRING"
  },
  {
    "mode": "NULLABLE",
    "name": "flashUsage",
    "type": "INT64"
  },
  {
    "mode": "REQUIRED",
    "name": "locale",
    "type": "STRING"
  },
  {
    "mode": "REQUIRED",
    "name": "os",
    "type": "STRING"
  },
  {
    "mode": "REQUIRED",
    "name": "osversion",
    "type": "STRING"
  },
  {
    "mode": "NULLABLE",
    "name": "profileDate",
    "type": "INT64"
  },
  {
    "fields": [
      {
        "mode": "REQUIRED",
        "name": "key",
        "type": "STRING"
      },
      {
        "mode": "REQUIRED",
        "name": "value",
        "type": "INT64"
      }
    ],
    "mode": "REPEATED",
    "name": "searches",
    "type": "RECORD"
  },
  {
    "mode": "REQUIRED",
    "name": "seq",
    "type": "INT64"
  },
  {
    "mode": "NULLABLE",
    "name": "sessions",
    "type": "INT64"
  },
  {
    "mode": "NULLABLE",
    "name": "showTrackerStatsShare",
    "type": "BOOL"
  },
  {
    "mode": "NULLABLE",
    "name": "tz",
    "type": "INT64"
  },
  {
    "mode": "REQUIRED",
    "name": "v",
    "type": "INT64"
  }
]
```

File: src/bigquery.rs [hunk 1]
```diff
@@ -150,6 +118,36 @@ impl From<ast::Tag> for Tag {
...
+    Root(Vec<Tag>),
+}
+
+impl From<ast::Tag> for Schema {
+    fn from(tag: ast::Tag) -> Self {
+        let mut bq_tag = Tag::from(tag.clone());
+        match *bq_tag.data_type {
+            Type::Record(_) if tag.is_array() || tag.is_map() => {
+                assert!(bq_tag.name.is_none());
+                bq_tag.name = Some("root".into());
```
Comment (fbertsch) [inline src/bigquery.rs]: nit: The name for the "default" root column should be statically defined

File: src/bigquery.rs [hunk 2]
```diff
@@ -150,6 +118,36 @@ impl From<ast::Tag> for Tag {
...
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+pub enum Schema {
+    Root(Vec<Tag>),
+}
+
+impl From<ast::Tag> for Schema {
+    fn from(tag: ast::Tag) -> Self {
+        let mut bq_tag = Tag::from(tag.clone());
+        match *bq_tag.data_type {
```
[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/bigquery.rs]: Unfortunately no, not without defining the `Clone` attribute on all of the related structures.

```rust
        match (&bq_tag.name, &*bq_tag.data_type) {
            (None, _) | (Some(_), Type::Atom(_)) => {
                assert!(bq_tag.name.is_none());
                bq_tag.name = Some("root".into());
                Schema::Root(vec![Box::new(bq_tag)])
            }
            (Some(_), Type::Record(record)) => {
                let mut vec: Vec<_> = record.fields.iter().collect();
                vec.sort_by_key(|(key, _)| key.to_string());
                let columns = vec.iter().map(|(_, v)| **v).collect();
                Schema::Root(columns)
            }
        }
```
I have a statement like this, but either the `records.fields.values` is coming from borrowed content, or `bq_tag` is partially used.

[the remark this answers is not in the document]
Comment (acmiyaguchi) [reply src/bigquery.rs]: I'll file an issue for this and leave it as it.

[the remark this answers is not in the document]
Comment (fbertsch) [reply src/bigquery.rs]: That's fine, I'll leave it up to you.

**Comment by acmiyaguchi** [comment]
> The "root" case of no named columns seems to only come up in testing.

If we had empty schemas in mps that weren't injected with metadata, we would see the `root` case more often. But you're right, this is an edge-case that we don't necessarily have to worry about. I'm fairly confident that we're doing the right thing if we were importing from an avro file that had a top-level primitive type though.
